### ISSUE TITLE
(aws-cdk/aws-s3): (bucket.grantRead doesn't do anything)

### ISSUE THREAD

**Issue by vpachipenta** [aws/aws-cdk#22060]
### Describe the bug

I created an S3 bucket in Stack1.  In Stack 2, referencing this bucket using `bucket.fromAttributes` methods.  In Stack2 I also created a CloudFrontDistribution and OriginAccessIdentity.  Then grant access to OriginAccessIdentity. After deploying these stacks, no updates made to Bucket policy.  Also no errors thrown during deployment.

### Expected Behavior

OriginAccessIdentity should be granted Read access on S3 Bucket.

### Current Behavior

It doesn't grant read access.

### Reproduction Steps

Code for Stack1
```
export default class StaticAssetsStack extends Stack {

    public readonly staticAssetsBucket : IBucket;
  
  constructor(parent: App, name: string, props: StaticAssetsStackProps) {
    super(parent, name, props);

    this.staticAssetsBucket = new Bucket(this, `AssetBucket`, {
        versioned: false,
        bucketName: `website-assets`.toLowerCase(),
        removalPolicy: RemovalPolicy.RETAIN,
        enforceSSL: true,
        encryption: BucketEncryption.S3_MANAGED,
      });
  }
}
```

Code for Stack2
```
export default class WebsiteStack extends Stack {
  constructor(parent: App, name: string, props: WebsiteStackProps) {
    super(parent, name, props);

     const staticAssetsBucket = Bucket.fromBucketAttributes(this, `staticAssetsBucket`, {
      bucketName: props.staticAssetsBucketName,
      region: props.staticAssetsBucketRegion
    });
   const originAccessIdentity =
      new OriginAccessIdentity(this, `${props.name}-OAI`);
   staticAssetsBucket.grantRead(originAccessIdentity);
   const s3Origin = new S3Origin(staticAssetsBucket, {
      originAccessIdentity,
    });
```

### CDK CLI Version

CDKv2

### Node.js Version

14.x

### OS

Linux

### Language

Typescript

**Comment by gshpychka** [comment]
This is currently the expected behavior - imported resources cannot be modified. To avoid this, pass the construct between the stacks the usual way - pass the construct instance as a constructor prop in Stack2

**Comment by kornicameister** [comment]
Well, that's not quite true, at least now all the way. One example can be found in my app, where I augmented legacy instance role with necessary permissions. So in other words; I had imported construct and was able to modify it.
Point taken; that was indirect modification via `ManagedPolicies` where role as attached. 

Just wanted to clarify that a bit; that some resources to support that while others don't.

**Comment by gshpychka** [comment]
> > pass the construct instance as a constructor prop in Stack2
> 
> This results in a circular reference error during cdk synth.
> 
> My observation is `grantRead` for an imported resource works if the target resource is a Lambda function. It doesn't do anything with OriginAccessIdentity. I believe, this is true, if the `grantRead` operation has to update S3 bucket policy.

Yes, `grantRead` modifies the identity policy of the grantee if possible, and the resource policy of the bucket if not. And since modifying an imported resource directly isn't possible, it's a no-op.

**Comment by peterwoodworth** [comment]
> Yes, grantRead modifies the identity policy of the grantee if possible, and the resource policy of the bucket if not. And since modifying an imported resource directly isn't possible, it's a no-op.

Thanks again for your help @gshpychka, you're exactly right. Perhaps we should throw a warning in the case where calling `grantRead` or similar methods doesn't result in anything

**Comment by gshpychka** [comment]
This is not a duplicate of #30953

**Comment by aemada-aws** [comment]
@gshpychka can you elaborate why is it not a duplicate of #30953? #30953 describes the same issue that `grantRead` on `OriginAccessIdentity` does nothing and AccessDenied is thrown. The example in that issue also uses an imported bucket.

**Comment by gshpychka** [comment]
> [@gshpychka](https://github.com/gshpychka) can you elaborate why is it not a duplicate of [#30953](https://github.com/aws/aws-cdk/issues/30953)? [#30953](https://github.com/aws/aws-cdk/issues/30953) describes the same issue that `grantRead` on `OriginAccessIdentity` does nothing and AccessDenied is thrown. The example in that issue also uses an imported bucket.

Because this issue was created 1.5 years before that one and is more general (the issue isn't specific to OAI or cloudfront-origins).


#30953 is the duplicate, not this one.

**Comment by aemada-aws** [comment]
@gshpychka Thanks for the explanation. The bot selects issues based on engagement count and the other issue has more engagement but since this issue is more general, I'm closing the other issue in favor of this more general one.
